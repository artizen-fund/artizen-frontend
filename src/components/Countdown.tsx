import { useState, useEffect } from 'react'
import styled from 'styled-components'
import { default as ReactCountdown } from 'react-countdown'
import moment from 'moment-timezone'
import { typography } from '@theme'
import { ARTIZEN_TIMEZONE } from '@lib'

export interface CountdownProps {
  date: Partial<any>
  onComplete?: () => void
}

interface IRendererProps {
  days: number
  hours: number
  minutes: number
  seconds: number
  completed: boolean
}

const Countdown = ({ date, onComplete }: CountdownProps) => {
  // date is save in PST so it needs to be converted to local machine date
  const newMomentDate = moment.tz(date, ARTIZEN_TIMEZONE).local().format()
  const [hasMounted, setHasMounted] = useState(false)

  useEffect(() => {
    // This component wrecks havoc with testing libraries.
    // This check fixes it.
    setHasMounted(typeof jest === 'undefined')
  }, [])

  const renderer = ({ days, hours, minutes, seconds, completed }: IRendererProps) => {
    if (!!completed) {
      if (typeof window === 'undefined') return
      window.location.reload()
    }
    return completed ? (
      <span key="resetting">resetting</span>
    ) : (
      <span key={`${days}.${hours}.${minutes}.${seconds}`}>
        {days}d: {hours}h: {minutes}m
      </span>
    )
  }

  return (
    <Wrapper>{hasMounted && <ReactCountdown autoStart {...{ date: newMomentDate, renderer, onComplete }} />}</Wrapper>
  )
}

const Wrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 10px;
  ${typography.title.l4}
`

export default Countdown
