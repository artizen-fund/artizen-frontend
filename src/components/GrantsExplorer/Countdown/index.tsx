import { useState, useEffect } from 'react'
import styled from 'styled-components'
import { default as ReactCountdown } from 'react-countdown'
import { Glyph } from '@components'

export interface CountdownProps {
  date: string
}

const Countdown = ({ date }: CountdownProps) => {
  const [hasMounted, setHasMounted] = useState(false)
  useEffect(() => setHasMounted(typeof jest === 'undefined'), [])
  return (
    <Wrapper>
      <Glyph glyph="countdown" />
      {hasMounted && <ReactCountdown {...{ date }} />}
    </Wrapper>
  )
}

const Wrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 10px;
`

export default Countdown
