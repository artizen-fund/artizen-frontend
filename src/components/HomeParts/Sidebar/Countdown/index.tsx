import styled from 'styled-components'
import { default as ReactCountdown } from 'react-countdown'
import { Icon } from '@components'
import { breakpoint, palette } from '@theme'
import { rgba } from '@lib'

export interface CountdownProps {
  date: string
}

const Countdown = ({ date }: CountdownProps) => (
  <Wrapper>
    <Icon>countdown</Icon>
    <ReactCountdown {...{ date }} />
  </Wrapper>
)

const Wrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 10px;
`

export default Countdown
