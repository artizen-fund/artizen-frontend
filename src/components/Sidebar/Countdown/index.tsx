import { useState, useEffect } from 'react'
import styled from 'styled-components'
import { default as ReactCountdown } from 'react-countdown'
import { Icon } from '@components'
import { isServer } from '@lib'

export interface CountdownProps {
  date: string
}

const Countdown = ({ date }: CountdownProps) => {
  /* note: This component causes obvious hydration errors.
   *       Utilizing useEffect to make sure the SSR and initial render are identical. */

  const [loaded, setLoaded] = useState(false)
  useEffect(() => {
    if (isServer()) return
    setLoaded(true)
  }, [])
  return (
    <Wrapper>
      <Icon>countdown</Icon>
      {loaded && <ReactCountdown {...{ date }} />}
    </Wrapper>
  )
}

const Wrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 10px;
`

export default Countdown
