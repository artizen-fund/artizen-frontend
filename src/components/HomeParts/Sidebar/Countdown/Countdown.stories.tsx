import Countdown, { CountdownProps } from './'

export default {
  title: 'page/home/Countdown',
  component: Countdown,
  argTypes: {},
}

export const CountdownComponent = (props: CountdownProps) => {
  return <Countdown {...props} />
}
