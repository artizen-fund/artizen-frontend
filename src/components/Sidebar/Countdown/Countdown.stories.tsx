import Countdown, { CountdownProps } from './'

const story = {
  title: 'page/home/Countdown',
  component: Countdown,
  argTypes: {},
}
export default story

export const CountdownComponent = (props: CountdownProps) => {
  return <Countdown {...props} />
}
