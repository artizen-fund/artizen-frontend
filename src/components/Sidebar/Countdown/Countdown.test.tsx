/* important todo: Snapshots don't like to work with this thing!
 *
 * This could be fixed -- in theory -- by feeding it a date value that is always
 * "one minute ago". Then every snapshot would record a time of five minutes ago.
 *
 * But say the time is 12:05:59:99, and the countdown timer runs a millisecond later.
 * It might record a snapshot of 12:00 or 12:01, and won't match a prior snapshot.
 *
 * "Maybe this component doesn't need snapshotsâ€¦" Agreed, but <Sidebar /> also breaks
 * because it includes this thing.
 *
 * I think we have to adjust this component to just not run on SSR (where the figure
 * would be perpetually inaccurate anyway) and on test scripts.
 */

import '@testing-library/jest-dom/extend-expect'
import { render } from '@testing-library/react'
import Countdown from './'

describe('Countdown', () => {
  it('renders Countdown unchanged', () => {
    const { container } = render(<Countdown date="2022-06-30T00:00:00" />)
    expect(container).toMatchSnapshot()
  })
})
