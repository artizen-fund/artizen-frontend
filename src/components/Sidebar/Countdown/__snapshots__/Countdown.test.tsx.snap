// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown renders Countdown unchanged 1`] = `
<div>
  <div
    class="sc-39826d64-0 ibsXOM"
  >
    <div
      class="sc-cf411747-2 cnUxdA"
    >
      <div
        class="sc-cf411747-0 hPyuHh"
        color="white"
      >
        <div
          class="sc-9eb58f28-0 gvNxsL"
          color="white"
        />
      </div>
    </div>
    <span>
      26
      :
      12
      :
      03
      :
      23
    </span>
  </div>
</div>
`;
